Defer reroute when starting shards

DaveCTurner opened the pull request:
Today we reroute the cluster as part of the process of starting a shard, which
runs at `URGENT` priority. In large clusters, rerouting may take some time to
complete, and this means that a mere trickle of shard-started events can cause
starvation for other, lower-priority, tasks that are pending on the master.

However, it isn't really necessary to perform a reroute when starting a shard,
as long as one occurs eventually. This commit removes the inline reroute from
the process of starting a shard and replaces it with a deferred one that runs
at `NORMAL` priority, avoiding starvation of higher-priority tasks.

This may improve some of the situations related to #42738 and #42105.

DaveCTurner committed:
Defer reroute when starting shards

Today we reroute the cluster as part of the process of starting a shard, which
runs at `URGENT` priority. In large clusters, rerouting may take some time to
complete, and this means that a mere trickle of shard-started events can cause
starvation for other, lower-priority, tasks that are pending on the master.

However, it isn't really necessary to perform a reroute when starting a shard,
as long as one occurs eventually. This commit removes the inline reroute from
the process of starting a shard and replaces it with a deferred one that runs
at `NORMAL` priority, avoiding starvation of higher-priority tasks.

This may improve some of the situations related to #42738 and #42105.

elasticmachine commented:
Pinging @elastic/es-distributed

DaveCTurner commented:
Note to reviewers: don't let the `+462 −515 Files changed 61` put you off too much. The largest part of this is adjusting all the test cases that use an `AllocationService` to start some shards, because with this change they also need to perform a separate reroute. Apart from that it's just a bit of plumbing to get the `RerouteService` in place, and boilerplate for the already-deprecated setting that can be removed in `master`.

original-brownbear commented:
@DaveCTurner seems there's a problem with the setting change on CI and it's probably just missing a warning exclusion in the rest tests:

```
java.lang.AssertionError: unexpected warning headers expected null, but was:<[299 Elasticsearch-8.0.0-SNAPSHOT-cf9953f "[cluster.routing.allocation.shard_state.reroute.priority] setting was deprecated in Elasticsearch and will be removed in a future release! See the breaking changes documentation for the next major version."]>
```

DaveCTurner committed:
Specific test case for followup priority setting

We cannot set the priority in all InternalTestClusters because the deprecation
warning makes some tests unhappy. This commit adds a specific test instead.

DaveCTurner commented:
Bit more complicated than that - I thought I could get away with something simple but got bitten by the deprecation logger, so had to add a specific test in 5fc2e7e.

DaveCTurner committed:
Checkstyle

original-brownbear commented:
Jenkins run elasticsearch-ci/2 (just the xpack license test failing)

original-brownbear reviewed (commented):
Code looks good, just one question.

Also, to me the change to `NORMAL` priority/not-inline-execution makes sense here but maybe Yannick should confirm as well :)
I wonder though if we really should hide this change as much as it's hidden now. Couldn't we add a note on this change stating that this might be a problem if your cluster is already badly configured (master overloaded + lots of shard fluctuation?) but is an optimization for a healthy setup? That way users affected negatively by this might be more likely to either understand/fix the problem or report back in a more meaningful way?

DaveCTurner committed:
Cluster state always changed here

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java:
```diff
@@ -589,6 +623,15 @@ public void onFailure(String source, Exception e) {
+
+        @Override
+        public void clusterStatePublished(ClusterChangedEvent clusterChangedEvent) {
+            if (clusterChangedEvent.previousState() != clusterChangedEvent.state()) {
```
You're quite right, this method doesn't behave like `clusterStateProcessed()` (which is always called).

DaveCTurner commented:
Sure. There's no way to express changes to the migration docs in a PR against `master` because the docs don't exist here, but I propose this in `7.x`:

```diff
diff --git a/docs/reference/migration/migrate_7_4.asciidoc b/docs/reference/migration/migrate_7_4.asciidoc
index ebfca7d25c1..63315f9fb65 100644
--- a/docs/reference/migration/migrate_7_4.asciidoc
+++ b/docs/reference/migration/migrate_7_4.asciidoc
@@ -67,4 +67,20 @@ unsupported on buckets created after September 30th 2020.
 Starting in version 7.4, a `+` in a URL will be encoded as `%2B` by all REST API functionality. Prior versions handled a `+` as a single space.
 If your application requires handling `+` as a single space you can return to the old behaviour by setting the system property
 `es.rest.url_plus_as_space` to `true`. Note that this behaviour is deprecated and setting this system property to `true` will cease
-to be supported in version 8.
\ No newline at end of file
+to be supported in version 8.
+
+[float]
+[[breaking_74_cluster_changes]]
+=== Cluster changes
+
+[float]
+==== Rerouting after starting a shard runs at lower priority
+
+After starting each shard the elected master node must perform a reroute to
+search for other shards that could be allocated. In particular, when creating
+an index it is this task that allocates the replicas once the primaries have
+started. In versions prior to 7.4 this task runs at priority `URGENT`, but
+starting in version 7.4 its priority is reduced to `NORMAL`. In a
+well-configured cluster this reduces the amount of work the master must do, but
+means that a cluster with a master that is overloaded with other tasks at
+`HIGH` or `URGENT` priority may take longer to allocate all replicas.
```

original-brownbear reviewed (approved):
LGTM :) thanks @DaveCTurner

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java:
```diff
@@ -589,6 +623,13 @@ public void onFailure(String source, Exception e) {
+        public void clusterStatePublished(ClusterChangedEvent clusterChangedEvent) {
+            rerouteService.reroute("reroute after starting shards", prioritySupplier.get(), ActionListener.wrap(
+                r -> logger.trace("reroute after starting shards succeeded"),
+                e -> logger.debug("reroute after starting shards failed", e)));
```
I wonder if we should log at DEBUG level here if `e instanceof FailedToCommitClusterStateException || e instanceof NotMasterException` but otherwise log at WARN level. Failing to do a delayed reroute sounds like a bug to me (unless there's a master failover).

DaveCTurner replied:
We already log at `ERROR` in the `BatchedRerouteService` on a failure other than no longer being a master (or `WARN` if we failed to submit the task). I think that's sufficient here.

ywelsch replied:
ok

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java:
```diff
@@ -109,7 +109,6 @@ public ClusterState applyStartedShards(ClusterState clusterState, List<ShardRout
         Collections.sort(startedShards, Comparator.comparing(ShardRouting::primary));
         applyStartedShards(allocation, startedShards);
         gatewayAllocator.applyStartedShards(allocation, startedShards);
-        reroute(allocation);
```
can you instead still call `assert RoutingNodes.assertShardStats(allocation.routingNodes());` here?

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java:
```diff
@@ -79,10 +81,23 @@
+     */
+    public static final Setting<Priority> FOLLOW_UP_REROUTE_PRIORITY_SETTING
+        = new Setting<>("cluster.routing.allocation.shard_state.reroute.priority", Priority.NORMAL.toString(),
+        name -> Priority.valueOf(name.toUpperCase(Locale.ROOT)),
```
I think we should disallow setting this to anything other than NORMAL, HIGH or URGENT.

ywelsch reviewed (commented):
I've left a few small comments, looking good o.w.

DaveCTurner committed:
Assert consistency of routing nodes

DaveCTurner committed:
Restrict setting only to reasonable priorities

DaveCTurner committed:
Merge branch 'master' into 2019-07-16-defer-reroute-when-starting-shards

ywelsch reviewed (approved):
LGTM
